fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricSoundsProvider.java

```diff
@@ -0,0 +1,117 @@
+import net.fabricmc.fabric.impl.datagen.SoundTypeBuilderImpl;
+
+/**
+ * Extend this class and implement {@link FabricSoundsProvider#generate}.
+ *
+ * <p>Register an instance of the class with {@link FabricDataGenerator.Pack#addProvider} in a {@link net.fabricmc.fabric.api.datagen.v1.DataGeneratorEntrypoint}.
+ *
+ * <p>Registered sound types will be appended to their own sounds.json in a namespace corresponding to
+ * the id of the sound event they are assigned to.
+ */
+public abstract class FabricSoundsProvider implements DataProvider {
+	private static final Codec<Map<String, SoundTypeBuilderImpl.SoundType>> CODEC = Codec.unboundedMap(Codec.STRING, SoundTypeBuilderImpl.SoundType.CODEC);
+	private final CompletableFuture<RegistryWrapper.WrapperLookup> registryLookupFuture;
+	private final DataOutput output;
+
+	public FabricSoundsProvider(DataOutput output, CompletableFuture<RegistryWrapper.WrapperLookup> registryLookupFuture) {
+		this.registryLookupFuture = registryLookupFuture;
+		this.output = output;
+	}
+
+	@Override
+	public CompletableFuture<?> run(DataWriter writer) {
+		return registryLookupFuture.thenCompose(lookup -> {
+			final Map<String, Map<String, SoundTypeBuilderImpl.SoundType>> data = new LinkedHashMap<>();
+			configure((id, builder) -> {
+				if (data.computeIfAbsent(id.getNamespace(), n -> new LinkedHashMap<>()).put(id.getPath(), ((SoundTypeBuilderImpl) builder).build()) != null) {
+					throw new IllegalStateException("Duplicate sound for event " + id);
+				}
+			});
+
+			return CompletableFuture.allOf(data.entrySet().stream().map(file -> {
+				Path outputPath = output.resolvePath(DataOutput.OutputType.RESOURCE_PACK).resolve(file.getKey() + "/sounds.json");
+				return DataProvider.writeCodecToPath(writer, lookup, CODEC, file.getValue(), outputPath);
+			}).toArray(CompletableFuture[]::new));
+		});
+	}
+
+	/**
+	 * Implement this method and then use {@link BiConsumer#accept} to register sound events to be data-generated.
+	 *
+	 * <p>Registered sound types will be appended to their own sounds.json in a namespace corresponding to
+	 * the id of the sound event they are assigned to.
+	 */
+	protected abstract void configure(SoundExporter exporter);
+
+	/**
+	 * A consumer used by {@link FabricSoundsProvider#configure}.
+	 */
+	@ApiStatus.NonExtendable
+	@FunctionalInterface
+	public interface SoundExporter {
+		/**
+		 * Adds a sound event.
+		 *
+		 * @param event   The sound event to get it id from.
+		 * @param builder The sound event details.
+		 */
+		default void add(SoundEvent event, SoundTypeBuilder builder) {
+			add(event.id(), builder);
+		}
+
+		/**
+		 * Adds a sound event.
+		 *
+		 * @param event   A registry entry for sound event to get it id from.
+		 * @param builder The sound event details.
+		 */
+		default void add(RegistryEntry<SoundEvent> event, SoundTypeBuilder builder) {
+			add(event.getKey().orElseThrow(() -> new NullPointerException("Sound event was not registered")).getValue(), builder);
```

Comment:
IllegalArgumentException, also change the message to "Direct (non-registered) sound event cannot be added"